{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f087bbd8",
   "metadata": {},
   "source": [
    "## 读取10x数据"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71bf1229",
   "metadata": {},
   "source": [
    "### sc.read_10_mtx()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fed19357",
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "adata = sc.read_10x_mtx(\n",
    "    \"data/filtered_gene_bc_matrices/hg19/\",  # the directory with the `.mtx` file\n",
    "    var_names=\"gene_symbols\",  # use gene symbols for the variable names (variables-axis index)\n",
    "    cache=True,  # write a cache file for faster subsequent reading\n",
    ")\n",
    "adata.var_names_make_unique()  # this is unnecessary if using `var_names='gene_ids'` in `sc.read_10x_mtx`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dd116231",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2700 × 32738\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f462cf8",
   "metadata": {},
   "source": [
    "### 源码解析"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "095c1f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scanpy.readwrite.py\n",
    "\n",
    "def read_10x_mtx(\n",
    "    path: Path | str,\n",
    "    *,\n",
    "    var_names: Literal[\"gene_symbols\", \"gene_ids\"] = \"gene_symbols\",\n",
    "    make_unique: bool = True,\n",
    "    cache: bool = False,\n",
    "    cache_compression: Literal[\"gzip\", \"lzf\"] | None | Empty = _empty,\n",
    "    gex_only: bool = True,\n",
    "    prefix: str | None = None,\n",
    ") -> AnnData:\n",
    "    \"\"\"Read 10x-Genomics-formatted mtx directory.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    path\n",
    "        Path to directory for `.mtx` and `.tsv` files,\n",
    "        e.g. './filtered_gene_bc_matrices/hg19/'.\n",
    "        文件路径，应包含3个文件：matrix.mtx, barcodes.tsv, genes.tsv\n",
    "    var_names\n",
    "        The variables index.\n",
    "        变量名称，使用基因符号名称，还是基因ID，默认使用基因符号\n",
    "    make_unique\n",
    "        Whether to make the variables index unique by appending '-1',\n",
    "        '-2' etc. or not.\n",
    "        在重复变量名称后面加-1、-2这样的后缀。\n",
    "    cache\n",
    "        If `False`, read from source, if `True`, read from fast 'h5ad' cache.\n",
    "        是否缓存原始文件到当前cache目录。\n",
    "    cache_compression\n",
    "        See the h5py :ref:`dataset_compression`.\n",
    "        (Default: `settings.cache_compression`)\n",
    "    gex_only\n",
    "        Only keep 'Gene Expression' data and ignore other feature types,\n",
    "        e.g. 'Antibody Capture', 'CRISPR Guide Capture', or 'Custom'\n",
    "        只保留基因表达值，其他值忽略\n",
    "    prefix\n",
    "        Any prefix before `matrix.mtx`, `genes.tsv` and `barcodes.tsv`. For instance,\n",
    "        if the files are named `patientA_matrix.mtx`, `patientA_genes.tsv` and\n",
    "        `patientA_barcodes.tsv` the prefix is `patientA_`.\n",
    "        (Default: no prefix)\n",
    "        数据文件名称前面是否有前缀\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    An :class:`~anndata.AnnData` object\n",
    "\n",
    "    \"\"\"\n",
    "    path = Path(path)\n",
    "    prefix = \"\" if prefix is None else prefix\n",
    "\n",
    "    # genes.tsv这样的格式为遗留格式\n",
    "    is_legacy = (path / f\"{prefix}genes.tsv\").is_file()\n",
    "    adata = _read_10x_mtx(\n",
    "        path,\n",
    "        var_names=var_names,\n",
    "        make_unique=make_unique,\n",
    "        cache=cache,\n",
    "        cache_compression=cache_compression,\n",
    "        prefix=prefix,\n",
    "        is_legacy=is_legacy,\n",
    "    )\n",
    "\n",
    "    # 遗留格式只有基因表达，直接返回。新格式，没有设置 gex_only，即保留所有数据，也直接返回。\n",
    "    if is_legacy or not gex_only:\n",
    "        return adata\n",
    "    \n",
    "    # 新格式，设置了 gex_only，只保留基因表达数据\n",
    "    gex_rows = adata.var[\"feature_types\"] == \"Gene Expression\"\n",
    "    return adata[:, gex_rows].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8048471",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _read_10x_mtx(\n",
    "    path: Path,\n",
    "    *,\n",
    "    var_names: Literal[\"gene_symbols\", \"gene_ids\"] = \"gene_symbols\",\n",
    "    make_unique: bool = True,\n",
    "    cache: bool = False,\n",
    "    cache_compression: Literal[\"gzip\", \"lzf\"] | None | Empty = _empty,\n",
    "    prefix: str = \"\",\n",
    "    is_legacy: bool,\n",
    ") -> AnnData:\n",
    "    \"\"\"Read mex from output from Cell Ranger v2- or v3+.\"\"\"\n",
    "\n",
    "    # 新格式文件以.gz结尾，遗留格式无.gz结尾\n",
    "    suffix = \"\" if is_legacy else \".gz\"\n",
    "\n",
    "    # 读取表达数据\n",
    "    adata = read(\n",
    "        path / f\"{prefix}matrix.mtx{suffix}\",\n",
    "        cache=cache,\n",
    "        cache_compression=cache_compression,\n",
    "    ).T  # transpose the data\n",
    "    # 转置，结果为：行是细胞，列是基因\n",
    "\n",
    "    # 读取基因表格：第1列 - ID，第2列 - 名称，第3列（新） - feature_types\n",
    "    # 旧表格叫 genes.tsv，新表格叫 features.tsv，不只基因，还含有其他特征\n",
    "    # 将基因名称或基因ID用作变量名称，即 adata.var 的索引\n",
    "    # 当用基因名称作为变量索引时，如果设置了去重，会将重复名称后面加-1、-2这样的数字，但名称列表的元素总数不会变\n",
    "    # 当用基因ID作为变量索引时，似乎不需要去重，或许是上游已经处理好了\n",
    "    genes = pd.read_csv(\n",
    "        path / f\"{prefix}{'genes' if is_legacy else 'features'}.tsv{suffix}\",\n",
    "        header=None,\n",
    "        sep=\"\\t\",\n",
    "    )\n",
    "    if var_names == \"gene_symbols\":\n",
    "        var_names_idx = pd.Index(genes[1].values)\n",
    "        if make_unique:\n",
    "            var_names_idx = anndata.utils.make_index_unique(var_names_idx)\n",
    "        adata.var_names = var_names_idx\n",
    "        adata.var[\"gene_ids\"] = genes[0].values\n",
    "    elif var_names == \"gene_ids\":\n",
    "        adata.var_names = genes[0].values\n",
    "        adata.var[\"gene_symbols\"] = genes[1].values\n",
    "    else:\n",
    "        msg = \"`var_names` needs to be 'gene_symbols' or 'gene_ids'\"\n",
    "        raise ValueError(msg)\n",
    "    if not is_legacy:\n",
    "        adata.var[\"feature_types\"] = genes[2].values\n",
    "    \n",
    "\n",
    "    # 读取barcodes，并将其赋给adata.obs_names，即adata.obs的索引\n",
    "    barcodes = pd.read_csv(path / f\"{prefix}barcodes.tsv{suffix}\", header=None)\n",
    "    adata.obs_names = barcodes[0].values\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6219f0c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 仅读取表达矩阵\n",
    "adata = sc.read('data/filtered_gene_bc_matrices/hg19/matrix.mtx')\n",
    "# 基因列表\n",
    "genes = pd.read_csv('data/filtered_gene_bc_matrices/hg19/genes.tsv', header=None, sep='\\t')\n",
    "# 细胞列表\n",
    "barcodes = pd.read_csv('data/filtered_gene_bc_matrices/hg19/barcodes.tsv', header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ac9ae7",
   "metadata": {},
   "source": [
    "### 探索AnnData结构\n",
    "\n",
    "obs, var 是一维注释（数据框）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80b072c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class AnnData(metaclass=utils.DeprecationMixinMeta):  # noqa: PLW1641\n",
    "    def _set_dim_index(self, value: pd.Index, attr: str):\n",
    "        # Assumes _prep_dim_index has been run\n",
    "        if self.is_view:\n",
    "            self._init_as_actual(self.copy())\n",
    "        getattr(self, attr).index = value\n",
    "        for v in getattr(self, f\"_{attr}m\").values():\n",
    "            if isinstance(v, pd.DataFrame):\n",
    "                v.index = value\n",
    "\n",
    "    @property\n",
    "    def obs(self) -> pd.DataFrame | Dataset2D:\n",
    "        \"\"\"One-dimensional annotation of observations (`pd.DataFrame`).\"\"\"\n",
    "        return self._obs\n",
    "\n",
    "    @obs.setter\n",
    "    def obs(self, value: pd.DataFrame | XDataset):\n",
    "        self._set_dim_df(value, \"obs\")\n",
    "\n",
    "    @obs.deleter\n",
    "    def obs(self):\n",
    "        self.obs = pd.DataFrame({}, index=self.obs_names)\n",
    "\n",
    "    @property\n",
    "    def obs_names(self) -> pd.Index:\n",
    "        \"\"\"Names of observations (alias for `.obs.index`).\"\"\"\n",
    "        return self.obs.index\n",
    "\n",
    "    @obs_names.setter\n",
    "    def obs_names(self, names: Sequence[str]):\n",
    "        names = self._prep_dim_index(names, \"obs\")\n",
    "        self._set_dim_index(names, \"obs\")\n",
    "\n",
    "    \n",
    "    @property\n",
    "    def var(self) -> pd.DataFrame | Dataset2D:\n",
    "        \"\"\"One-dimensional annotation of variables/ features (`pd.DataFrame`).\"\"\"\n",
    "        return self._var\n",
    "\n",
    "    @var.setter\n",
    "    def var(self, value: pd.DataFrame | XDataset):\n",
    "        self._set_dim_df(value, \"var\")\n",
    "\n",
    "    @var.deleter\n",
    "    def var(self):\n",
    "        self.var = pd.DataFrame({}, index=self.var_names)\n",
    "\n",
    "    @property\n",
    "    def var_names(self) -> pd.Index:\n",
    "        \"\"\"Names of variables (alias for `.var.index`).\"\"\"\n",
    "        return self.var.index\n",
    "\n",
    "    @var_names.setter\n",
    "    def var_names(self, names: Sequence[str]):\n",
    "        names = self._prep_dim_index(names, \"var\")\n",
    "        self._set_dim_index(names, \"var\")\n",
    "\n",
    "    \n",
    "    def var_names_make_unique(self, join: str = \"-\"):\n",
    "        # Important to go through the setter so obsm dataframes are updated too\n",
    "        self.var_names = utils.make_index_unique(self.var.index, join)\n",
    "\n",
    "    def obs_names_make_unique(self, join: str = \"-\"):\n",
    "        # Important to go through the setter so obsm dataframes are updated too\n",
    "        self.obs_names = utils.make_index_unique(self.obs.index, join)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c5f063ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2700 × 32738\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names_make_unique()\n",
    "adata"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
